src/sabyenc3.cc

```diff
@@ -571,6 +577,147 @@ PyObject* decode_usenet_chunks(PyObject* self, PyObject* Py_input_list) {
     return retval;
 }
 
+PyObject* decode_buffer(PyObject* self, PyObject* args) {
+    // The input/output PyObjects
+    (void)self;
+    PyObject *retval = NULL;
+    PyBytesObject *Py_bytes_obj = NULL;
+    PyObject *Py_output_filename = NULL;
+    PyObject *Py_output_crc = NULL;
+    int data_length;
+
+    // Used buffers
+    char *cur_char = NULL;
+    char *start_loc = NULL;
+    char *end_loc = NULL;
+    char *dest_loc = NULL;
+    uint32_t crc = 0;
+    uint32_t crc_yenc = 0;
+    size_t yenc_data_length;
+    size_t output_len;
+    const char* crc_pos;
+
+    // Parse input
+    if (!PyArg_ParseTuple(args, "Si:decode", &Py_bytes_obj, &data_length)) {
+        return NULL;
+    }
+    dest_loc = cur_char = (char *)Py_bytes_obj->ob_sval;
+    end_loc = dest_loc + data_length;
+
+    // Check for valid size
+    if(data_length <= 0 || data_length > Py_SIZE(Py_bytes_obj)) {
+        PyErr_SetString(PyExc_ValueError, "Invalid data length");
+        retval = NULL;
+        goto finish;
+    }
+
+    // Lift the GIL
+    Py_BEGIN_ALLOW_THREADS;
+
+    /*
+     ANALYZE HEADER
+     Always in the same format, e.g.:
+
+     =ybegin part=41 line=128 size=49152000 name=90E2Sdvsmds0801dvsmds90E.part06.rar
+     =ypart begin=15360001 end=15744000
+
+     But we only care about the filename
+    */
+    // Start of header
+    start_loc = my_memstr(cur_char, end_loc - cur_char, "=ybegin", 1);
+    if (!start_loc) {
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Invalid yEnc header");
+        retval = NULL;
+        goto finish;
+    }
+
+    // Find start of the filename
+    start_loc = my_memstr(start_loc, end_loc - start_loc, " name=", 1);
+    if (!start_loc) {
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Could not find yEnc filename");
+        retval = NULL;
+        goto finish;
+    }
+
+    // Extract filename
+    cur_char = start_loc;
+    for (; *cur_char != SABYENC_LF && *cur_char != SABYENC_CR && *cur_char != SABYENC_ZERO; cur_char++);
```

Comment:
Possibly for safety this should also check that cur_char < end_loc (same o nthe next one)